internal/proxy/meta_cache.go

```diff
@@ -68,15 +77,23 @@ type partitionInfo struct {
 	createdUtcTimestamp uint64
 }
 
+type credentialInfo struct {
+	username string
+	password string // hashed password
+}
+
 // make sure MetaCache implements Cache.
 var _ Cache = (*MetaCache)(nil)
 
 // MetaCache implements Cache, provides collection meta cache based on internal RootCoord
 type MetaCache struct {
 	client types.RootCoord
 
-	collInfo map[string]*collectionInfo
-	mu       sync.RWMutex
+	collInfo         map[string]*collectionInfo
+	credMap          map[string]*credentialInfo // cache for credential, lazy load
+	credUsernameList []string                   // no need initialize when NewMetaCache
```

Comment:
Why not use the key set of credMap instead credUsernameList?

Reply:
credMap is part of credentials, it is loaded as required.
credUsernameList is used to tell if we need turning basic-auth on, and also for showing the user list.